### PR TITLE
GO-2046 | GO-2044 | GO-1890 Multiple template apply support

### PR DESCRIPTION
**Pull request by KirillSto**
---

- [x] I understand that contributing to this repository will require me to agree with the [CLA](https://github.com/anyproto/.github/blob/main/docs/CLA.md)

---

### Description

This PR adds multiple template applying on object creation.
Additionally fixed a bug with title changing on template application and added ChangeSnapshot support in export archive unpacker

### What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI

### Related Tickets & Documents

https://linear.app/anytype/issue/GO-2044/objectapplytemplate-doesnt-change-object-name
https://linear.app/anytype/issue/GO-2046/research-ability-to-change-template-multiple-types
https://linear.app/anytype/issue/GO-1890/make-archiveconverter-capable-to-handle-snapshot-pbs

### Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

### Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 [tech-docs](https://github.com/anyproto/tech-docs)
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: core/block/service.go [hunk 1]
```diff
@@ -1014,35 +1020,33 @@ func (s *Service) DoFileNonLock(id string, apply func(b file.File) error) error
...
+		ts, err := s.StateFromTemplate(templateID, "")
 		if err != nil {
 			return err
 		}
-		ts.SetRootId(contextId)
+		ts.SetRootId(contextID)
 		ts.SetParent(orig)
 
-		fromLayout, _ := orig.Layout()
+		fromLayout, _ := ts.Layout()
```
Comment (deff7) [inline core/block/service.go:1033]: maybe use basic.SetLayoutInState here?

File: docs/Flow.md [hunk 2]
```diff
@@ -20,4 +20,44 @@ Need to redesign this approach the way to handle every possible change of object
 - Creation of bookmark
 - ...
 
-As now backlinks are gotten from the object store, all these changes need to be done after store redesign, so mechanism of getting links of object and saving it both to store and to backlinks relation has same enter point.
\ No newline at end of file
+As now backlinks are gotten from the object store, all these changes need to be done after store redesign, so mechanism of getting links of object and saving it both to store and to backlinks relation has same enter point.
+
+## Internal Flags
```
Comment (deff7) [inline docs/Flow.md:25]: Maybe make it as comments in code? This separate documentation will decay over time, much faster than commented code

File: core/block/editor/stext/text.go [hunk 3]
```diff
@@ -274,10 +274,19 @@ func (t *textImpl) newSetTextState(blockId string, ctx session.Context) *state.S
 
 func (t *textImpl) flushSetTextState(_ smartblock.ApplyInfo) error {
 	if t.lastSetTextState != nil {
+		applyFlags := []smartblock.ApplyFlag{smartblock.NoHooks}
+		textChanged, err := t.isLastTextBlockChanged()
```
Comment (deff7) [inline core/block/editor/stext/text.go:278]: as I mentioned in PM, you should encapsulate working with KeepInternalFlags as much as possible, like

```
applyFlags = []smatblock.ApplyFlag{smartblock.NoHooks}
if t.shouldKeepInternalFlags() {
   ...
}
```

take care to name those functions to be easily searchable via `Find in files`

Comment (KirillSto) [reply core/block/editor/stext/text.go]: Added `shouldKeepInternalFlags` function in textImpl and basic

File: core/block/service.go (hunk 1 above)
Comment (KirillSto) [reply core/block/service.go]: Done!

File: core/block/editor/stext/text.go [hunk 4]
```diff
@@ -462,6 +472,29 @@ func (t *textImpl) TurnInto(ctx session.Context, style model.BlockContentTextSty
...
 
+func (t *textImpl) isLastTextBlockChanged() (bool, error) {
+	if t.lastSetTextState == nil || t.lastSetTextId == "" {
+		return true, fmt.Errorf("last state about text block is not saved")
+	}
+	newTextBlock, err := getText(t.lastSetTextState, t.lastSetTextId)
+	if err != nil {
+		return true, err
+	}
+	oldTextBlock := t.Pick(t.lastSetTextId)
```
Comment (deff7) [inline core/block/editor/stext/text.go:483]: seems like you could use `t.lastSetTextState.PickOrigin` (or get parent state and pick block from it without using helper PickOrigin), looks little bit confusing that we compare blocks from what is looks like different states. Sure they have parent-child relationship (t is parent of lastSetTextState), but more transparency won't hurt

**Review by deff7 — CHANGES_REQUESTED** [review]

File: core/block/object/objectcreator/creator.go [hunk 5]
```diff
@@ -92,15 +92,15 @@ func (c *Creator) Name() (name string) {
...
+	StateFromTemplate(templateID, name string) (st *state.State, err error)
 	CreateTreeObject(ctx context.Context, spaceID string, tp coresb.SmartBlockType, initFunc smartblock.InitFunc) (sb smartblock.SmartBlock, err error)
 	DeriveTreeObjectWithUniqueKey(ctx context.Context, spaceID string, key domain.UniqueKey, initFunc smartblock.InitFunc) (sb smartblock.SmartBlock, err error)
 	TemplateClone(spaceID string, id string) (templateID string, err error)
 }
 
 func (c *Creator) CreateSmartBlockFromTemplate(ctx context.Context, spaceID string, sbType coresb.SmartBlockType, objectTypeKeys []domain.TypeKey, details *types.Struct, templateID string) (id string, newDetails *types.Struct, err error) {
 	var createState *state.State
-	if templateID != "" {
+	if templateID != "" && templateID != block.BlankTemplateID {
```
[the remark this answers is not in the document]
Comment (KirillSto) [reply core/block/object/objectcreator/creator.go:104]: As clients use **blank** constant in fileds of Sets' and Collections' views, **templateId="balnk"** can (for now I think even DO) pass in CreateObject command. So in this case Middle should do nothing and just create empty state

That's why btw **BlockTemplateID** constant is exported

File: core/block/service.go [hunk 6]
```diff
@@ -64,11 +65,13 @@ import (
 const (
 	CName           = "block-service"
 	linkObjectShare = "anytype://object/share?"
+	BlankTemplateID = "blank"
```
[the remark this answers is not in the document]
Comment (KirillSto) [reply core/block/service.go:70]: This constant is used in **objectcreator** package, as clients could pass this value via **ObjectCreate** and **BlockLinkCreateWithObject** requests

File: core/block/object/objectcreator/creator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (deff7) [reply core/block/object/objectcreator/creator.go:104]: client logic must do this, you can do
```
if templateID == "blank" {
 templateID = ""
}

return c.CreateSmartBlockFromTemplate(ctx, spaceID, sbType, []domain.TypeKey{objectTypeKey}, details, templateID)

```

File: core/block/service.go [hunk 7]
```diff
@@ -1014,35 +1020,33 @@ func (s *Service) DoFileNonLock(id string, apply func(b file.File) error) error
...
 		}
-		ts.SetRootId(contextId)
+		ts.SetRootId(contextID)
 		ts.SetParent(orig)
 
-		fromLayout, _ := orig.Layout()
-		if toLayout, ok := orig.Layout(); ok {
-			if err := s.layoutConverter.Convert(ts, fromLayout, toLayout); err != nil {
+		layout, found := orig.Layout()
+		if commonOperations, ok := b.(basic.CommonOperations); found && ok {
```
[the remark this answers is not in the document]
Comment (KirillSto) [reply core/block/service.go]: Fixed

File: core/block/editor/stext/text.go (hunk 4 above)
Comment (KirillSto) [reply core/block/editor/stext/text.go]: Yeah, used PickOrigin for retrieving old state - need to fix tests though

File: core/block/editor/stext/text_test.go [hunk 8]
```diff
@@ -545,3 +545,64 @@ func TestTextImpl_TurnInto(t *testing.T) {
...
+func TestTextImpl_isLastTextBlockChanged(t *testing.T) {
+	text := "text"
+	blockID := "test"
+
+	t.Run("text inside last text block is changed", func(t *testing.T) {
+		// given
+		sb := smarttest.New(blockID)
+		sb.AddBlock(newTextBlock(blockID, text))
+		tb := NewText(sb, nil, nil)
+		tb.(*textImpl).lastSetTextId = blockID
```
[the remark this answers is not in the document]
Comment (KirillSto) [reply core/block/editor/stext/text_test.go]: Done!

File: core/block/object/objectcreator/creator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (KirillSto) [reply core/block/object/objectcreator/creator.go]: Done!

**Review by deff7 — APPROVED** [review]

**Pull request merged**